src/core/data/transform.cc

```diff
@@ -301,6 +316,19 @@ void Transpose::print(std::ostream& out) const
 
 int32_t Transpose::target_ndim(int32_t source_ndim) const { return source_ndim; }
 
+void Transpose::return_promoted_dims(std::vector<int32_t>& dims) const
+{
+  for (size_t i = 0; i < axes_.size(); i++) {
+    if (i != axes_[i]) {
+      auto finder_dims = std::find(dims.begin(), dims.end(), (i));
+      if (finder_dims != dims.end()) {
+        auto finder_axes   = std::find(axes_.begin(), axes_.end(), (i));
+        dims[*finder_dims] = *finder_axes;
+      }
+    }
+  }
```

Comment:
```suggestion
  // i should be added to X.tranpose(axes).promoted iff axes[i] is in X.promoted
  // e.g. X.promoted = [0] => X.transpose((1,2,0)).promoted = [2]
  for (auto& promoted : dims) {
    auto finder = std::find(axes_.begin(), axes_.end(), promoted);
#ifdef DEBUG_LEGATE
    assert(finder != axes_.end());
#endif
    promoted = finder - axes_.begin();
  }
```

The current code has the potential to blow up, because you're making multiple passes over the promoted dims, so the second pass has the potential to see the partial result from the first pass. I can't immediately spot a problem, but I think it's more safe and understandable if you iterate over the promoted dims instead.